File: src/test/java/ar/com/flow/bankaccount/transaction/TransactionTest.java
package ar.com.flow.bankaccount.transaction;

import ar.com.flow.bankaccount.BankAccount;
import ar.com.flow.bankaccount.transaction.receipt.DebitStamper;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import static ar.com.flow.bankaccount.BankAccountAssert.assertThat;
import static ar.com.flow.bankaccount.TestObjects.createSavingsAccountFor;
import static ar.com.flow.bankaccount.TestObjects.francisco;
import static ar.com.flow.money.TestObjects.dollars10;
import static ar.com.flow.money.TestObjects.dollars100;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public class TransactionTest {
    private BankAccount debitAccount;
    private Algorithm debit;
    private Algorithm creditMock;

    @BeforeEach
    void setUp() {
        debitAccount = createSavingsAccountFor(francisco, dollars100);
        debit = new Algorithm(debitAccount, new DebitStamper(debitAccount, Action.Transfer));
        creditMock = mock(Algorithm.class);
    }

    @Test
    void rollbackAfterFirstStep() {
        when(creditMock.execute(dollars10)).thenThrow(new RuntimeException());

        Transaction.builder()
                .step(debit)
                .step(creditMock)
                .amount(dollars10)
                .build()
                .execute();

        assertThat(debitAccount).keepsInitialBalance();
    }
}

File: src/test/kotlin/ar/com/flow/bankaccount/transaction/TransactionTest.kt
package ar.com.flow.bankaccount.transaction

import ar.com.flow.bankaccount.BankAccount
import ar.com.flow.bankaccount.BankAccountAssert.Companion.assertThat
import ar.com.flow.bankaccount.TestObjects.createSavingsAccountFor
import ar.com.flow.bankaccount.TestObjects.francisco
import ar.com.flow.bankaccount.transaction.receipt.Action
import ar.com.flow.bankaccount.transaction.receipt.DebitPrint
import ar.com.flow.bankaccount.transaction.receipt.DebitScratch
import ar.com.flow.bankaccount.transaction.steps.Step
import ar.com.flow.money.TestObjects.dollars10
import ar.com.flow.money.TestObjects.dollars100
import io.mockk.every
import io.mockk.mockk
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test

class TransactionTest {
    private lateinit var debitAccount: BankAccount
    private lateinit var debit: Step

    @BeforeEach
    fun setUp() {
        debitAccount = createSavingsAccountFor(francisco, dollars100)
        debit = Step(
            debitAccount,
            DebitPrint(debitAccount, Action.Transfer),
            DebitScratch(debitAccount, Action.Transfer)
        )
    }

    @Test
    fun rollbackAfterFirstStep() {
        val creditMock = mockk<Step>()

        every { creditMock.execute(dollars10) } throws RuntimeException()

        Builder()
            .steps(debit, creditMock)
            .amount(dollars10)
            .build()
            .execute()

        assertThat(debitAccount).keepsInitialBalance()
    }
}
